### DISCUSSION TITLE
Mid-show commercial breaks?

### DISCUSSION THREAD

**Discussion by CaliMiller** [chrisbenincasa/tunarr#292] [Ideas]
Could something like this be harnessed to add commercial breaks to films and TV series?

https://www.scenedetect.com/

**Comment by chrisbenincasa** [comment]
Wow this is very interesting. I thought about this as well way back when first hooking up DTV. This is definitely something I will look into once we stabilize a bit, thanks for the tip.

**Comment by avalonMMXXII** [comment]
Mid commercial breaks, even if it does not detect when the scene turns to a blank screen (as long as it does commercial breaks every set number of minutes) ...it is much better than watching 5-15 minutes of commercials after and episode or movie finishes.

**Comment by chrisbenincasa** [comment]
Couldn't agree more about the annoyance of 5+ mins of commercials between shows. When padding works, it works pretty well. And when it doesn't it's rather annoying.

We have two separate ideas here: the mid-show breaks, which IMO is hard to get "right" in terms of feel. We could do something quick and dirty, which maybe would be interesting, but we'd need to test it out. Anything more sophisticated would be a significant effort. That said, I love the idea.

For a simple way for "better" commercial fillers, I think we could capture this by implementing ETV "filler kinds" (#177 ). Instead of start-time-based filler, I like the idea of saying "show three 30-second" fillers between each media item. Or, "fill 90 seconds of time with filler" between each media item. The different supports kinds are here: :https://ersatztv.org/docs/user-guide/filler/#filler-kinds. I think Tunarr would benefit greatly from these.

**Reply by avalonMMXXII** [reply]
Thanks, yes it could used as an option as you mention for people that want to use it....this would still pad time but also add a break during the content being watched. Even if it is not perfect, its better than nothing if it can be done.

**Comment by CaliMiller** [comment]
I know this might be more work for you as a developer, but what if there was an option for the end user to choose between the two ideas?

whynotbothgirl.gif

**Reply by chrisbenincasa** [reply]
Oh yea I was not suggesting we'd have one or the other. But most likely mid-show breaks would be a potentially enormous undertaking. I think it's worth researching and playing around with, but I'm not sure about timeframe or priority for it at the moment.

**Comment by binarygeek119** [comment]
what about using chapter marks. i use fileflows to add chaper mark to all my video files and it does a good job like scenedetect. it be nice to have it build in but for now you could have a external tool do the work. but tunarr would need to use chapter mark to insert commercials.

**Comment by chrisbenincasa** [comment]
Yeah ffmpeg scene detect / black-screen detect could get us a lot of the way here for items that do not have chapter metadata available. Anecdotally, most of my TV shows do not have chapter data, but most of my movies do (especially Blu-Rays). I'm not sure how common it is for folks to go in and add the chapters for everything. 

There are _many_ tools out there other than ffmpeg that can do this as well (e.g. PySceneDetect). I'd like to do an evaluation of all of them then come up with a phased plan on how to implement this with some basic heuristics around what is an expected TV viewing experience. We also have to be mindful of packaging external tools given that we also distribute Tunarr as a "standalone" binary.

I did a really rough test with the `ffmpeg` `blackdetect` filter on a few TV shows in my library and it worked surprisingly well, especially for older shows. I think with some logic around the raw data we could emulate something pretty close to TV (e.g. 30 min shows get 2 breaks, rough near 7.5m and 15m; 45m shows get 3 breaks roughly near 10m, 22m, 34m). 

If we go the route of performing analysis there's a lot of other complexities to solve for too. A few things that come to mind: 1. when do we analyze? how to keep track of analysis version (if we improve later, we probably want to reanalyze)? performing this all via network inputs to ffmpeg can be quite slow compared to reading the videos from the filesystem... but folks don't always have access to the media files they can stream, etc.

Long-term I'd love to have a lot of options to customize this along with several methods, i.e. chapters, fallback to analysis, etc.

**Comment by binarygeek119** [comment]
as i said fileflows add them but what i should have said is it does it automatically. i have it run on all my video files. it converts then all to h.265 in a mkv with 5.1 acc sound with chapters added. its a set and forget once you have it dialed in.

**Reply by chrisbenincasa** [reply]
Sounds great! I think we'll have to fine a free and open source solution to integrate with Tunarr, though.

**Comment by dude22072** [comment]
Dropping this while it's on my mind, Plex should return the chapters contained within a video file (at least the XML stuff does), so if someone wants to do chapter breaks and is using plex as the source, the chapters could be grabbed that way instead of having to access the file itself. Unsure if Jellyfin has a way to do this.

**Comment by starkjames** [comment]
Adding my vote of support for this idea.  I'm currently using Tunarr to simulate 90s cable TV, thus sticking to :30 and :00 time blocks.  That means 7-9 minutes of commercials at the end of each "30 minute" episode and 14-18 minutes of commercials at the end of each "60 minute" episode.  The first is tolerable, but the second not so much.  Scene detection is an interesting way to handle it, though I don't think every break fades to black as some have suggested.

**Comment by Kevin-999** [comment]
I thought I saw some mention of this when I first learned of Tunarr. Just spotted another potential path for implementing this.

https://www.reddit.com/r/PleX/comments/1kzc7rw/update_on_commercialbreaker_toonamistyle_anime/
